changelog.d/20074_protobuf_decoder.breaking.md

```diff
@@ -0,0 +1,2 @@
+protobuf decoder will no longer set fields that are not set in the incoming byte stream. This to
```

Comment:
@flavioc is this breaking change necessary for this PR? Could we split into two?

@jszwedko the change itself looks reasonable to me. But since it's a breaking change, I want to hear your thoughts as well.

Reply:
I'm not sure I understand the behavior change here exactly. @flavioc could you maybe provide an example to illustrate?

Reply:
The difference is when we convert a proto message into a VRL value. In the old code we do:

```
prost_reflect::Value::Message(v) => {
            let mut obj_map = ObjectMap::new();
            for field_desc in v.descriptor().fields() {
                let field_value = v.get_field(&field_desc);
                let out = to_vrl(field_value.as_ref(), Some(&field_desc))?;
                obj_map.insert(field_desc.name().into(), out);
            }
            vrl::value::Value::from(obj_map)
        }
```

In the new code, we do:

```
prost_reflect::Value::Message(v) => {
            let mut obj_map = ObjectMap::new();
            for field_desc in v.descriptor().fields() {
                if v.has_field(&field_desc) {
                    let field_value = v.get_field(&field_desc);
                    let out = proto_to_value(field_value.as_ref(), Some(&field_desc))?;
                    obj_map.insert(field_desc.name().into(), out);
                }
            }
            Value::from(obj_map)
        }
```

This means that if a value is not set in the protocol buffer, we will not set the field in the VRL map.
This can be seen in the test deserialize_empty_buffer where before the value had the `name` field as `""` but now the value will be an empty dictionary. This ensures that if we encode the resulting value we get the same bytes that we received, which would not be the case before. It is also easier to check if fields are unset.

Reply:
Ah, gotcha, I see. Thanks for explaining. I agree that that sounds reasonable. I appreciate the call out as a breaking change.